### PR TITLE
Add derived equals for other types within Teleport.

### PR DESCRIPTION
**Pull request by mdwn**
Derived equals have been added for other types within Teleport with two notable exceptions:

- AccessLists, as there is a bug with goderive when determining equivalence for the ResourceHeader.
- Users, as goderive's equal plugin does not support interface{} types, which is used in the underlying UserV2 type generated by protoc.

Note: I've not added tests here, as we're primarily relying on goderive's `equal` plugin here. For the cases it supports, it seems to be fairly well tested. Let me know if you disagree and would like to see some custom tests.

### CODE REVIEW

File: api/types/header/derived.gen.go
[the remark this answers is not in the document]
Comment (mdwn) [reply api/types/header/derived.gen.go:1]: These are generated from the process shown in this PR: https://github.com/gravitational/teleport/pull/38981

[the remark this answers is not in the document]
Comment (Tener) [reply api/types/header/derived.gen.go:1]: Neat!

**Review by Tener — APPROVED** [review]

**Comment by zmb3** [comment]
How do these generated files get updated when the specs change? I would expect to see a go:generate pragma somewhere.

It would also be nice if the generated code contained the standard comment indicating that it is generated code: https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

**Comment by mdwn** [comment]
> How do these generated files get updated when the specs change? I would expect to see a go:generate pragma somewhere.
> 
> It would also be nice if the generated code contained the standard comment indicating that it is generated code: https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

These are updated by running `make derive` and a linting job will fail if they're out of date.

I'll look into that pragma and the comment.

**Comment by mdwn** [comment]
> How do these generated files get updated when the specs change? I would expect to see a go:generate pragma somewhere.
> 
> It would also be nice if the generated code contained the standard comment indicating that it is generated code: https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

So the generated code (which resides in `derived.gen.go`) does have this comment. The other code (the actual `IsEqual` functions are technically manual, but `goderive` generates the code labeled with `deriveTeleportEqual` when it detects it in the code base.

**Review by nklaassen — APPROVED** [review]

**Pull request merged**